### ISSUE TITLE
Discussion regarding future pull request implementing custom socket types

### ISSUE THREAD

**Issue by lucaszanella** [xia-chu/ZLMediaKit#428]
Hi, I like ZLMediaKit a lot, it's very well written and useful, but there's one thing that's missing: the hability to specify my own `Socket` instance derived from `ZLToolKit`, instead of letting the `MediaPlayer` create its own. 

Why do I need this? Well, I'm creating a type of virtual socket that sends everything through OpenVPN, instead of using the Posix or Windows API for opening a socket. So I need a custom `Socket` to be passed to `MediaPlayer`.

I've read though the code of both `ZLMediaKit` and `ZLToolKit`. On `ZLToolKit`, I intend to template the `TcpClient` class like this:

 ```
template<class SocketType = Socket>
class TcpClient : public std::enable_shared_from_this<TcpClient>, public SocketHelper{
public:
TcpClient(const EventPoller::Ptr &poller = nullptr);
TcpClient(SocketType::Ptr socket, const EventPoller::Ptr &poller = nullptr);
```

with this, no code modification is needed for people already using the library because `SocketType` defaults to the original `Socket`. Then, I'd simply template every class that depends on `TcpClient`, making them default to `Socket` also, until I get to `MediaPlayer`.

With this I can create a MediaPlayer and pass my custom socket instance:

` MediaPlayer::Ptr player(new MediaPlayer(myCustomSocketInstance));`

Do you have any comments and **would you accept this pull request**? 

Thank you very much!

**Comment by xia-chu** [comment]
In fact, I have implemented similar functions, but I have implemented it by deriving the `TcpClient` class. You can view the code of `TcpClientWithSSL`  and the `WebSocketClient`, which are implemented by template.

If you have a better way to achieve it, then I certainly welcome your pull request.

**Comment by lucaszanella** [comment]
Taking a closer look at `TcpClientWithSSL`, now I see that it overrides some of the `Socket` functions andproxies their calls to the `SSL_Box`. There are functions that are not virtual such as `alive` which calls `socket.rawFD()`. Looks like you rely on `TcpClientWithSSL` to use the native socket to start a connection, then after connection you pass the socket calls (`send`, etc) to the `SSL_Box`.

In my case there could be no system socket, it should be a totally virtual socket. So one option would be to make all of the `Socket` functions virtual so I can create something like `TcpClientWithSSL` that overrides all of them. 

There's still the problem of creating this special socket. On your code you create a `TcpClientWithSSL` based on the url type. However, I don't want to have special URLs. I simply want to create a virtual socket to work with any URL. Also, I need to create it before creating the `MediaPlayer`, so for sure I need to create a constructor on `MediaPlayer` that receives this special socket.

Since `MediaPlayer` stores a **pointer** to `PlayerBase`, I can take advantage of dynamic dyspatching of all the socket functions. 

So looks like I need to create a constructor for `MediaPlayer` that accepts a `PlayerBase::Ptr`

	MediaPlayer(const EventPoller::Ptr &poller = nullptr);
	MediaPlayer(PlayerBase::Ptr socket, const EventPoller::Ptr &poller = nullptr) {
              _socket = socket;
        }

then

```
    _parser = PlayerBase::createPlayer(_poller,strUrl, _socket);
```

Then on `createPlayer(const EventPoller::Ptr &poller,const string &strUrl, PlayerBase::Ptr _socket)`:

    if (strcasecmp("rtsps",prefix.data()) == 0) {
                if (_socket)
                    return PlayerBase::Ptr(_socket, releasePlayer);
                else
		    return PlayerBase::Ptr(new TcpClientWithSSL<RtspPlayerImp>(poller), releasePlayer);
	}

	if (strcasecmp("rtsp",prefix.data()) == 0) {
                if (_socket)
                    return PlayerBase::Ptr(_socket,releasePlayer);
                else
		    return PlayerBase::Ptr(new RtspPlayerImp(poller),releasePlayer);
	}

	if (strcasecmp("rtmps",prefix.data()) == 0) {
                if (_socket)
                    return PlayerBase::Ptr(_socket,releasePlayer);                
                else
		        return PlayerBase::Ptr(new TcpClientWithSSL<RtmpPlayerImp>(poller, _socket),releasePlayer);
	}

	if (strcasecmp("rtmp",prefix.data()) == 0) {
                if (_socket)
		        return PlayerBase::Ptr(_socket,releasePlayer);                
                else
		        return PlayerBase::Ptr(new RtmpPlayerImp(poller),releasePlayer);
	}

Where `_socket` would be a class that does what `TcpClientWithSSL` does, but instead of proxying the calls to an `SSL_Box` it'd proxy to whatever I implemented it to do. Remember that I'd have to make all of the functions on `Socket` virtual so I can erase some of them to not call anything on traditional system sockets.

What do you think of this solution? Looks like it eliminates the need for templates, which is great since otherwise I'd have to template lots of things. I'm also not sure if I got everything correct but if it works with `TcpClientWithSSL` then it should work with something similar to it but that overrides all functions of its parent (`Socket`).

**Comment by xia-chu** [comment]
You can try it. I think the workload may be a little too heavy, but if you do it well, I think I will be happy to accept your pull request.

**Comment by lucaszanella** [comment]
Ok I started working on it yesterday. I did minimal changes, you'll be happy. But there's some technical difficulty that I arrived at. I may need to pass my own `EventPoller` to `MediaPlayer`, which is not a problem since this was already supported. For that however I'll need to possibly make all the `EventPoller` functions virtual so I can override with my own. I cannot rely on `epoll` because my socket is all virtual, there are no file descriptors. 

It looks like when a socket has data, the `EventPoller` calls its `onRead`:

```
int result = _poller->addEvent(pSock->rawFd(), Event_Read | Event_Error | Event_Write, [weakSelf,weakSock,isUdp](int event) {
        auto strongSelf = weakSelf.lock();
        auto strongSock = weakSock.lock();
        if(!strongSelf || !strongSock) {
            return;
        }
        if (event & Event_Error) {
            strongSelf->onError(strongSock);
            return;
        }
        if (event & Event_Read) {
            strongSelf->onRead(strongSock,isUdp);
        }
        if (event & Event_Write) {
            strongSelf->onWriteAble(strongSock);
        }
    });
```
I don't even need an `EventPoller`, because I'm going to use just 'virtual sockets'.  My `Socket` class already has its own event looper that gives me back my packets. Is it safe for me to simple ignore the `EventPoller` passed to `MediaPlayer` which is passed to `Socket` and call the `_readCB` of the `Socket` from my `Socket` class event loop? Then the `EventPoller` will still be used for other things like timers. My main concern is about which thread should call the `_readCB`.

For what I've been able to find, the `EventPoller` passed to `MediaPlayer` is passed down up to the `TcpClient`,  which uses it for a timer and then for its sockets.

Does the EventPoller really needs to read the socket and call `_readCb`?

**Comment by xia-chu** [comment]
If you no longer trigger callbacks in the eventpoller thread, there may be a thread safety problem.
But you can implement the code first. I can make some changes when I merge it.

**Comment by lucaszanella** [comment]
Ok. I don't know if it's ready because I can't test if I dont trigger `_readCb` on the `Socket`. Since as you said, we must call things from the `EventPoller`, then there should be a way to post tasks to the `EventPoller` to be executed as soon as possible. But the problem is that the `EventPoller` waits with `epoll_wait`, which can only be untriggered forcefully if we write data to one of the sockets, but I don't find this solution good. Ideally there should be a way on both `epoll_wait` AND a `std::condition_variable`, and return whenever one of them has data.

Please take a look at my changes:

https://github.com/xia-chu/ZLToolKit/compare/master...lucaszanella:master

https://github.com/xia-chu/ZLMediaKit/compare/master...lucaszanella:master

(I added a simply development .devcontainer for VSCode but I can take it off if you prefer, it's just a way to edit the project and build it very fast, as when you open this project with VSCode, it detects the devcontainer.json and automatically builds the entire environment for you)

The `virtual Socket* clone(const EventPoller::Ptr &poller);` is needed because I noticed that on RtspPlayer you open additional sockets for Rtmp and Rtp, but I need these too to be sockets of my type, not new sockets of type `Socket`, so I created this as a way to clone the socket easily when needed

**Comment by lucaszanella** [comment]
Hi, just to inform that I found some bugs on the commits I sent in the last comment. Now I updated them so now the library works if I pass no custom socket, and when I pass my custom socket it's working up to the point where the EventPoller should call the onRead callback. Previously if I passed no custom socket then a problem on the clone function was getting the program to crash. So if you tested and got undefined behaviour and can test again now it should work.

Looks like there is a way to execute tasks on the poller thread with `doDelayTask`, I just created a new method that does that without the timer:

https://github.com/lucaszanella/ZLToolKit/blob/5aedb19ee792f572e1b54cae045114d39ac92411/src/Poller/EventPoller.cpp#L70

I don't know if it's right, I'm testing now and getting a crash but it might be on my code, gonna try to fix it.

**Comment by lucaszanella** [comment]
By the way, it crashes when I call 

`typedef function<void(const Buffer::Ptr &buf, struct sockaddr *addr , int addr_len)> onReadCB;`

with the first received RTSP message from the camera.

I've put the buffer received from my virtual socket inside a `toolkit::Buffer` like this:

```
class MergedBuffer : public toolkit::Buffer
{
public:
    MergedBuffer(std::shared_ptr<smoltcp::Buffer> smolBuffer) : smolBuffer(smolBuffer){};
    virtual ~MergedBuffer(){
        std::cout << "going to destruct MergedBuffer" << std::endl;
    };
    //返回数据长度
    virtual char *data() const
    {
        return reinterpret_cast<char*>(smolBuffer->getData());
    }
    virtual uint32_t size() const
    {
        return smolBuffer->len;
    }

private:
    std::shared_ptr<smoltcp::Buffer> smolBuffer;
};
```

and I pass this buffer like this:
```

_poller->doTaskOnPollerThread([mergedBuffer, &readCB, cIpAddress]() mutable {
                int len;
                struct sockaddr_in addr = cipaddress_to_ip(cIpAddress, &len);
                //I'v tried both ways, both crash
                //readCB(mergedBuffer, (struct sockaddr *)&addr, len);
                readCB(mergedBuffer, nullptr, 0);
            });
```

but I get an `free(): invalid pointer`. Simply commenting `readCB` shows that no crashes happens, so my buffer is able to delete itself. The backtrace:
```

* thread #11, name = 'vpn_rtsp_client', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff771718b libc.so.6`raise + 203
    frame #1: 0x00007ffff76f6859 libc.so.6`abort + 299
    frame #2: 0x00007ffff77613ee libc.so.6`___lldb_unnamed_symbol331$$libc.so.6 + 670
    frame #3: 0x00007ffff776947c libc.so.6`___lldb_unnamed_symbol399$$libc.so.6 + 28
    frame #4: 0x00007ffff776acac libc.so.6`___lldb_unnamed_symbol412$$libc.so.6 + 748
    frame #5: 0x00005555555f5ca0 vpn_rtsp_client_cli`__gnu_cxx::new_allocator<std::_Sp_counted_ptr_inplace<smoltcp::Buffer, std::allocator<smoltcp::Buffer>, (__gnu_cxx::_Lock_policy)2> >::deallocate(std::_Sp_counted_ptr_inplace<smoltcp::Buffer, std::allocator<smoltcp::Buffer>, (__gnu_cxx::_Lock_policy)2>*, unsigned long) + 36
    frame #6: 0x00005555555f536d vpn_rtsp_client_cli`std::allocator_traits<std::allocator<std::_Sp_counted_ptr_inplace<smoltcp::Buffer, std::allocator<smoltcp::Buffer>, (__gnu_cxx::_Lock_policy)2> > >::deallocate(std::allocator<std::_Sp_counted_ptr_inplace<smoltcp::Buffer, std::allocator<smoltcp::Buffer>, (__gnu_cxx::_Lock_policy)2> >&, std::_Sp_counted_ptr_inplace<smoltcp::Buffer, std::allocator<smoltcp::Buffer>, (__gnu_cxx::_Lock_policy)2>*, unsigned long) + 47
    frame #7: 0x00005555555f4182 vpn_rtsp_client_cli`std::__allocated_ptr<std::allocator<std::_Sp_counted_ptr_inplace<smoltcp::Buffer, std::allocator<smoltcp::Buffer>, (__gnu_cxx::_Lock_policy)2> > >::~__allocated_ptr() + 60
    frame #8: 0x00005555555f6d1c vpn_rtsp_client_cli`std::_Sp_counted_ptr_inplace<smoltcp::Buffer, std::allocator<smoltcp::Buffer>, (__gnu_cxx::_Lock_policy)2>::_M_destroy() + 112
    frame #9: 0x00005555555ddec4 vpn_rtsp_client_cli`std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() + 124
    frame #10: 0x00005555555dcecf vpn_rtsp_client_cli`std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() + 43
    frame #11: 0x00005555555e9792 vpn_rtsp_client_cli`std::__shared_ptr<smoltcp::Buffer, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() + 32
    frame #12: 0x00005555555e97b2 vpn_rtsp_client_cli`std::shared_ptr<smoltcp::Buffer>::~shared_ptr() + 28
```

Am I missing something about how the `onReadCB` works? Should I pass a buffer that never self destructs? I see that on `Socket.cpp` you use a buffer that indeed is never deleted. But since I already receive a ready buffer from my virtual socket, I thought it's better to use it, not to copy it again to this new buffer.

**Comment by lucaszanella** [comment]
Ok the crash was because I didn't put the `\0` at the end of the buffer. Maybe it shouldn't crash but simply throw an exception. If by any means the bufer does not have the `\0` then a vulnerability could be exploited. My code makes sure it never misses the `\0` but who knows.

So, it looks like everything is working and the `EventPoller` doesn't even need to be modified, because instead of calling `doTaskOnPollerThread` I can simply call `async_first` directly so I took that off. I'll test a little more and send you a pull request. 

I'm pretty happy with the modifications, they're very minimal:

https://github.com/xia-chu/ZLToolKit/compare/master...lucaszanella:master

https://github.com/xia-chu/ZLMediaKit/compare/master...lucaszanella:master

**Comment by xia-chu** [comment]
I will check your code next, you can submit a pull request first , thank you very much for your support!

**Comment by xia-chu** [comment]
I'm sorry that I'm busy recently and didn't deal with your issue in time.
I might have time to look at it this weekend，but I can't guarantee it。

**Comment by lucaszanella** [comment]
Thank you very much! I think you don't need to do anything right now, I'll prepare the pull request. I'm testing things first. The native `Socket` works normally even after my updates (that is, calling `new MediaPlayer()` without passing a `Socket` will simply work with the default socket). I also made it work with my custom socket by doing `new MediaPlayer(myCustomSocket)` and it worked but I get RTP timeout after some time, gonna figure out why before sending the pull request. Maybe something more needs to be done on ZLMediaKit/ZLToolKit or maybe it's just a problem on my custom `Socket`. We'll see.

**Comment by lucaszanella** [comment]
Found the error! I was doing

```
_poller->async_first([&readCB, bb, this]() mutable {
                LOCK_GUARD(_mutexEvent);
                readCB(buffer, nullptr, 0);
            },
```

to post packets from the socket to the TCP client (RTSP client). Doing this only worked for a while... Turns out I was posting the readCB events to the end of the queue or something like that, se they'd arrive out of order to the client. Changing from `async_first` to `async` worked, now they arrive in order! Nothing's wrong with ZLMediaKit or my library, it was just the glue that was wrong.

Gonna send you a pull request this week! Thank you :)

**Comment by xia-chu** [comment]
`async_first` means to execute the pushed task with the highest priority， so the order of execution is out of order

**Comment by xia-chu** [comment]
This feature has been implemented , please refer to #468.

**Comment by lucaszanella** [comment]
currently my setup is broken, gonna fix it and test as soon as possible, thank you very much!
